v4 (#1)
Skn0tt/adapters: .changeset/ten-cougars-pretend.md

---
'@astrojs/netlify': major
---

# Netlify Adapter v4 simplifies static + SSR deployments

This update is a complete overhaul of the Netlify adapter.
It simplifies the user-facing config, and resolves a number of bugs along the way.

Here's what changes:

## Middleware without configuration

In v3, [Middleware](https://docs.astro.build/en/guides/middleware/) was deployed to the SSR Function by default.
Enabling the `edgeMiddleware` option would change this behaviour to deploy it using Netlify Edge Functions instead, improving
page speed and enabling the middleware to run on prerendered sites.

In v4, this is the default setting.
If your Astro site makes use of Middleware,
it's deployed to an Edge Function where it runs super close to your users and is also applied to static sites.
You can use this to implement customization or authorization logic!

**Action Required:**

> If your site is using Middleware, and you had `edgeMiddleware` enabled, nothing changes for you - feel free to remove the config field, though.

Reviewer: [pebble] Maybe something more direct for the last part of the sentence?

```suggestion
If your site is using Middleware, and you had `edgeMiddleware` enabled, nothing changes for you. You are free to remove the config field, as it is no longer necessary.
```

Note that I'm assuming the config field is no longer necessary based on the way this is written?

> If your site is using Middleware, and you didn't have `edgeMiddleware` enabled, your Middleware will now be deployed to an Edge Function (probably no change needed) and start running on prerendered pages (check if this requires any changes to your middleware logic).

Reviewer: [pebble] I feel like the `probably no change needed` brings up questions so I would get rid of it. I would also move the last part to its own sentence, outside of the parentheses.

```suggestion
If your site is using Middleware, and you didn't have `edgeMiddleware` enabled, your Middleware will now be deployed to an Edge Function and start running on prerendered pages. Check if this requires any changes to your middleware logic.
```

However, if we do say `Check if this requires any changes to your middleware logic` it would be nice to say something as to how they can check it? Or link to another section that has the information on how to check for changes?

Author: > it would be nice to say something as to how they can check it?

They should look at all prerendered pages, and check if they still work as intended now that the middleware applies to them. Will add another line about that 👍

Author: how do you like https://github.com/Skn0tt/adapters/pull/1/commits/f52963b00fcb7d18e5facc073232c3bd0c2f406b?

## Netlify Context is automatically available via Locals

In v3, you could use `netlify-edge-middleware.ts` to inject data from the Netlify context into your Astro locals.
In v4, this file is no longer needed because the Netlify context is automatically made available via `Astro.locals.netlify.context`.

> You can use this to access information about the user (like geolocation or IP address), your Netlify site (like deploy ID) or the request (like its request ID or the CDN region it's served from).

Reviewer: [sand] To help add some clarity

```suggestion
You can use this context to access information about the user (like geolocation or IP address), your Netlify site (like deploy ID) or the request (like its request ID or the CDN region it's served from).
```

**Action Required:**

> If you didn't have a `netlify-edge-middleware.ts` or `netlify-edge-middleware.js` file, you're good.

Reviewer: [pebble] Maybe it would help to have a little more info at the end? Like "If you don't have either of these files, you can continue" or something similar? (I'm not sure what the logical "next step" would be here.)

I would also move this to be below the 'Remove the file' content.

Author: addressed with https://github.com/Skn0tt/adapters/pull/1/commits/7d53619e9111e3f85fd646bb04728b140b45ce69

Remove that file.
In your codebase, change all usage of locals injected through that file to use `Astro.locals.netlify.context` instead.

## On-Demand Builders are no longer supported

On-Demand Builders (ODB) allows SSR-Rendered pages to be cached using a Time to Live (TTL) strategy.

> ODB continues to be supported by Netlify, but has since been superceded by the much more powerful

Reviewer 2: ```suggestion
ODB continues to be supported by Netlify, but has since been superseded by the much more powerful
```

> ODB continues to be supported by Netlify, but has since been superseded by the much more powerful

Reviewer: [sand] Simplifying language

```suggestion
ODB continues to be supported by Netlify, but has since been replaced by the much more powerful
```

[Fine-Grained Cache Control](https://www.netlify.com/blog/swr-and-fine-grained-cache-control).

In v3, you could deploy your SSR-Rendered Astro pages to ODBs by enabling the `builders` config option,
and then specifying the TTL on a per-page basis.
In v4, you can use the cache-control headers instead:

```diff lang="astro"
---
// src/pages/index.astro
import Layout from '../components/Layout.astro';

- if (import.meta.env.PROD) {
-   Astro.locals.runtime.setBuildersTtl(45)
- }
+ Astro.response.headers.set('CDN-Cache-Control', "public, max-age=45, must-revalidate")
---
```

**Action Required:**

> If you weren't using ODBs, you're good.

Reviewer: [pebble] Same comments as the last `Action Required` comment. I would change the ending to give some kind of action instead of just "you're good". I would also move the sentence to be below the "remove" option content.

Reviewer: As a side note, another option for this section and the previous one would be to remove the "you're good" lines altogether. I noticed the following sections just say to remove the option without a failsafe of "but if you don't have this, you're okay". Maybe we can just do that here?

I assume that users would know whether or not they are using ODBs /middleware, so maybe they don't need these extra notes?

Author: i've removed the "you're good" lines in https://github.com/Skn0tt/adapters/pull/1/commits/7d53619e9111e3f85fd646bb04728b140b45ce69. agree, they're not necessary!

---
'@astrojs/netlify': major
---

# Netlify Adapter v4 simplifies static + SSR deployments

This update is a complete overhaul of the Netlify adapter.
It simplifies the user-facing config, and resolves a number of bugs along the way.

Here's what changes:

## Middleware without configuration

In v3, [Middleware](https://docs.astro.build/en/guides/middleware/) was deployed to the SSR Function by default.
Enabling the `edgeMiddleware` option would change this behaviour to deploy it using Netlify Edge Functions instead, improving
page speed and enabling the middleware to run on prerendered sites.

In v4, this is the default setting.
If your Astro site makes use of Middleware,
it's deployed to an Edge Function where it runs super close to your users and is also applied to static sites.
You can use this to implement customization or authorization logic!

**Action Required:**
If your site is using Middleware, and you had `edgeMiddleware` enabled, nothing changes for you. You are free to remove the config field, as it is no longer necessary.

> If your site is using Middleware, and you didn't have `edgeMiddleware` enabled, your Middleware will now be deployed to an Edge Function and start running on prerendered pages. Go through your prerendered pages, and check if this requires any changes to the logic in your middleware.

Reviewer: [pebbles] Some comma fixes, one small word fix.

```suggestion
If your site is using Middleware and you had `edgeMiddleware` enabled, nothing changes for you. You are free to remove the config field, as it is no longer necessary.
If your site is using Middleware and you didn't have `edgeMiddleware` enabled, your Middleware will now be deployed to an Edge Function and start running on prerendered pages. Go through your prerendered pages to check if this requires any changes to the logic in your middleware.
```

## Netlify Context is automatically available via Locals

In v3, you could use `netlify-edge-middleware.ts` to inject data from the Netlify context into your Astro locals.
In v4, this file is no longer needed because the Netlify context is automatically made available via `Astro.locals.netlify.context`.
You can use this context to access information about the user (like geolocation or IP address), your Netlify site (like deploy ID) or the request (like its request ID or the CDN region it's served from).

**Action Required:**

> Remove the `netlify-edge-middleware.ts` or ``netlify-edge-middleware.js` file.

Reviewer: [pebble] I think there's an extra backtick here that can be removed

```suggestion
Remove the `netlify-edge-middleware.ts` or `netlify-edge-middleware.js` file.
```

Remove the `netlify-edge-middleware.ts` or ``netlify-edge-middleware.js` file.
In your codebase, change all usage of locals injected through that file to use `Astro.locals.netlify.context` instead.

### Image CDN

> v4 of this adapter integrates your Astro site with Netlifys [Image CDN](https://docs.netlify.com/image-cdn/overview/) feature.

Reviewer 2: ```suggestion
v4 of this adapter integrates your Astro site with Netlify [Image CDN](https://docs.netlify.com/image-cdn/overview/).
```
